Nos/Service/Relay/RelaySubscriptionManager.swift

```diff
@@ -114,58 +122,90 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
         }
         return staleSubscriptions
     }
+    
+    // MARK: - Socket Management
+    
+    /// Opens sockets to any relays that we have an open subscription for that don't already have a socket.
+    func openSockets() {
+        var relayAddresses = Set<URL>()
+        all.forEach { relayAddresses.insert($0.relayAddress) }
 
-    func addSocket(for relayAddress: URL) -> WebSocket? {
-        guard !sockets.contains(where: { $0.request.url == relayAddress }) else {
-            return nil
-        }
-        
-        if let priorError = errored[relayAddress],
-            priorError.nextRetry > Date.now {
-            // This socket has errored recently and it isn't yet time to retry again.
-            return nil
+        for relayAddress in relayAddresses {
+            var connection: WebSocketConnection
+            if let existingConnection = socketConnections[relayAddress] {
+                connection = existingConnection
+            } else {
+                connection = createSocket(for: relayAddress)
+            }
+            
+            switch connection.state {
+            case .errored(let error):
+                if error.nextRetry > Date.now {
+                    return
+                } else {
+                    fallthrough
+                }
+            case .disconnected:
+                connection.socket.connect()
+                connection.state = .connecting
+            case .connected, .connecting:
+                return
+            }
         }
-        
+    }
+    
+    /// Creates a WebSocketConnection for a relay. This is not idempotent - make sure it's called only once for 
+    /// each relay.
```

Comment:
What happens if it's called more than once per relay? It looks like we'd replace the old connection with a new one in the dictionary, maybe? Could we make it idempotent, or is there a reason not to?

Reply:
If it's called more than once then we will leave dangling WebSockets open that won't close until the app is closed. 

I think I could make it idempotent though with some refactoring to `openSockets()` - I'll try that.

Reply:
@mplorentz did you end up doing this? I approved and am good with it either way -- just want to know.